src/driver-sinowealth.c

```diff
@@ -0,0 +1,475 @@
+	rc = ratbag_hidraw_set_feature_report(device, SINOWEALTH_REPORT_ID_CMD, cmd, sizeof(cmd));
+	if(rc != sizeof(cmd)) {
+		log_error(device->ratbag, "Error while sending read config command: %d\n", rc);
+		return -1;
+	}
+
+	rc = ratbag_hidraw_get_feature_report(device, SINOWEALTH_REPORT_ID_CONFIG,
+					      (uint8_t*) config, SINOWEALTH_CONFIG_SIZE);
+	/* The GET_FEATURE report length has to be 520, but the actual data returned is less */
+	if (rc < sizeof(*config) - sizeof(config->padding)) {
+		log_error(device->ratbag, "Could not read device configuration: %d\n", rc);
+		return -1;
+	}
+
+	/* TODO */
+	ratbag_profile_set_report_rate_list(profile, &hz, 1);
+	ratbag_profile_set_report_rate(profile, hz);
+
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if(config->config & SINOWEALTH_XY_INDEPENDENT) {
+			resolution->dpi_x = sinowealth_raw_to_dpi(config->dpi[resolution->index * 2]);
+			resolution->dpi_y = sinowealth_raw_to_dpi(config->dpi[resolution->index * 2 + 1]);
+		} else {
+			resolution->dpi_x = sinowealth_raw_to_dpi(config->dpi[resolution->index]);
+			resolution->dpi_y = resolution->dpi_x;
+		}
+		if (config->dpi_enabled & (1<<resolution->index)) {
+			/* DPI step is disabled, fake it by setting DPI to 0 */
+			resolution->dpi_x = 0;
+			resolution->dpi_y = 0;
+		}
+		resolution->is_active = resolution->index == config->active_dpi - 1;
+		resolution->is_default = resolution->is_active;
+	}
+
+	/* Body lighting */
+	led = ratbag_profile_get_led(profile, 0);
+	switch (config->rgb_effect) {
+	case RGB_OFF:
+		led->mode = RATBAG_LED_OFF;
+		break;
+	case RGB_SINGLE:
+		led->mode = RATBAG_LED_ON;
+		led->color = sinowealth_rbg_to_color(config->single_color);
+		led->brightness = sinowealth_rgb_mode_to_brightness(config->single_mode);
+		break;
+	case RGB_GLORIOUS:
+	case RGB_BREATHING:
+	case RGB_BREATHING7:
+	case RGB_TAIL:
+	case RGB_RAVE:
+	case RGB_WAVE:
+		led->mode = RATBAG_LED_CYCLE;
+		led->brightness = sinowealth_rgb_mode_to_brightness(config->glorious_mode);
+		led->ms = sinowealth_rgb_mode_to_speed(config->glorious_mode);
+		break;
+	case RGB_BREATHING1:
+		led->mode = RATBAG_LED_BREATHING;
+		led->color = sinowealth_rbg_to_color(config->breathing1_color);
+		led->brightness = sinowealth_rgb_mode_to_brightness(config->breathing1_mode);
+		led->ms = sinowealth_rgb_mode_to_speed(config->breathing1_mode);
+		break;
+	}
+	ratbag_led_unref(led);
+
+	/* DPI indicator LED */
+	for (int i = 1; i < SINOWEALTH_NUM_DPIS + 1; i++) {
+		led = ratbag_profile_get_led(profile, i);
+		led->mode = RATBAG_LED_ON;
+		led->color = sinowealth_raw_to_color(config->dpi_color[i - 1]);
+		ratbag_led_unref(led);
+	}
+
+	profile->is_active = true;
+
+	return 0;
+}
+
+static void
+sinowealth_init_profile(struct ratbag_device *device)
+{
+	struct ratbag_profile *profile;
+	struct ratbag_resolution *resolution;
+	struct ratbag_led *led;
+	int num_dpis = (SINOWEALTH_DPI_MAX - SINOWEALTH_DPI_MIN) / SINOWEALTH_DPI_STEP + 1 + 1;
```

Comment:
`+1 + 1`? I've paged most of ratbag out by now but this looks odd to me.

Reply:
"+ 1" because (max-min)/step counts like an half-open interval, but we want both min and max included, another "+ 1" because we have "0" as an extra special value to signal a DPI step being disabled -- I'll add an comment and make it "+2".